### PR TITLE
Physical ports

### PR DESCRIPTION
**Pull request by david-deboer**
Added a "hookup" feature to look at the "physical" ports in the system.

### CODE REVIEW

File: hera_mc/cm_handling.py [hunk 1]
```diff
@@ -39,7 +39,7 @@ def __init__(self, hpn, rev, at_date):
         self.connections = None  # This is the PartConnectionDossierEntry class
         self.geo = None  # This is the geo_location.GeoLocation class
 
-    def __repr__(self):
+    def __repr__(self):  # pragma: no cover
```
Comment (bhazelton) [inline hera_mc/cm_handling.py]: this should be coverable. There's an example starting on line 35 of `test_server_status.py`.

File: hera_mc/cm_handling.py [hunk 2]
```diff
@@ -382,6 +392,34 @@ def show_parts(self, part_dossier):
             table_data.append(part_dossier[hpnr].table_entry_row(columns))
         print('\n' + tabulate(table_data, headers=headers, tablefmt='orgtbl') + '\n')
 
+    def get_physical_connections(self, at_date=None):
+        """
+        Finds and returns a list of "physical (@)" connections.
```
Comment (bhazelton) [inline hera_mc/cm_handling.py]: I'm not sure what this line in the docstring means. Specifically what does the '(@)' mean? And maybe a little more clarity around what "physical" means (why is it in quotes?)

Comment (david-deboer) [reply hera_mc/cm_handling.py]: I made the requested changes and threw in a few more tests within cm_handling

File: hera_mc/tests/test_connections.py [hunk 3]
```diff
@@ -96,6 +97,9 @@ def test_get_dossier(self):
         self.assertTrue(len(x) == 0)
         x = self.h.get_part_connection_dossier('test_part2', 'active', 'all', at_date=old_time, exact_match=True)
         self.assertTrue(len(x) == 0)
+        z = {}
+        z['tst'] = cm_handling.PartConnectionDossierEntry('test_part1', 'active', 'all', at_date='now')
+        self.h.show_connections(z)
```
Comment (bhazelton) [inline hera_mc/tests/test_connections.py]: Testing print statements needs to be done by capturing the output and making sure that what it prints out is what you expect.

Comment (bhazelton) [reply hera_mc/tests/test_connections.py]: https://stackoverflow.com/questions/4219717/how-to-assert-output-with-nosetest-unittest-in-python

Comment (plaplant) [reply hera_mc/tests/test_connections.py]: Example in hera_cal: https://github.com/HERA-Team/hera_cal/blob/master/hera_cal/tests/test_utils.py#L429

It uses the context manager defined at the top of that test file.

File: hera_mc/tests/test_connections.py [hunk 4]
```diff
@@ -76,24 +92,43 @@ def test_update_new(self):
...
-        self.h.show_connections(located)
+        with captured_output() as (out, err):
+            self.h.show_connections(located)
+        self.assertTrue('new_test_part_up:A | up_and_out  | down_and_in | new_test_part_down:A' in out.getvalue().strip())
+        with captured_output() as (out, err):
+            self.h.show_connections(located, verbosity=1)
+        self.assertTrue('new_test_part_up:A | new_test_part_down:A' in out.getvalue().strip())
+        with captured_output() as (out, err):
+            self.h.show_connections(located, verbosity=2)
+        print(out.getvalue().strip())
```
Comment (bhazelton) [inline hera_mc/tests/test_connections.py]: Do we still want this print statement?

File: hera_mc/tests/test_parts.py [hunk 5]
```diff
@@ -122,6 +148,7 @@ def test_get_part_types(self):
         at_date = self.now
         a = self.h.get_part_types('all', at_date)
         self.assertTrue('terminals' in a['feed']['output_ports'])
+        self.h.show_part_types()
```
Comment (bhazelton) [inline hera_mc/tests/test_parts.py]: One more place that needs the context manager, I think.

File: hera_mc/tests/test_sys_handling.py [hunk 6]
```diff
@@ -51,14 +51,14 @@ def test_other_hookup(self):
         self.assertEqual(hookup.cached_hookup_dict['A23:H'].hookup['e'][0].upstream_part, 'HH23')
         hu = hookup.get_hookup('cached', 'H', 'pol', force_new=False)
         hu = hookup.get_hookup('A23,A23', 'H', 'pol', force_new=False)
+        hookup.cached_hookup_dict = None
+        hookup.determine_hookup_cache_to_use()
```
Comment (bhazelton) [inline hera_mc/tests/test_sys_handling.py]: what does this line do? Shouldn't there be an assert statement after this?

File: hera_mc/cm_handling.py [hunk 7]
```diff
@@ -382,6 +392,38 @@ def show_parts(self, part_dossier):
             table_data.append(part_dossier[hpnr].table_entry_row(columns))
         print('\n' + tabulate(table_data, headers=headers, tablefmt='orgtbl') + '\n')
 
+    def get_physical_connections(self, at_date=None):
+        """
+        Finds and returns a list of "physical" connections, as opposed to "hookup" connections.
+            In this context "hookup" refers to all signal path connections that uniquely determine
+            the path from station to correlator input.
+            "Physical" refers to other connections that we wish to track, such as power or rack location.
+            The leading character of physical ports is '@'.
```
Comment (bhazelton) [inline hera_mc/cm_handling.py]: Thanks, this is great.

File: scripts/parts.py [hunk 8]
```diff
@@ -42,6 +42,7 @@
             conn_info:  provide a summary of connections to given part/rev/port
             rev_info:  provide a summary of revisions of given part/rev
             types:  provide a summary of part types
+            physical:  shows the "physical" (@) port connections
```
Comment (bhazelton) [inline scripts/parts.py]: Please expand this a little.

**Pull request merged**